Mathlib/NumberTheory/Ostrowski.lean

```diff
@@ -55,4 +57,163 @@ lemma equiv_on_Nat_iff_equiv : (∃ c : ℝ, 0 < c ∧ (∀ n : ℕ , (f n)^c =
+    have ha :=
+      le_of_not_lt <| not_and.1 ((hmin a).mt hap.not_le) (map_pos_of_ne_zero f (mod_cast ha₀))
+    have hb :=
+      le_of_not_lt <| not_and.1 ((hmin b).mt hbp.not_le) (map_pos_of_ne_zero f (mod_cast hb₀))
+    rw [Nat.cast_mul, map_mul] at hp1
+    exact ((one_le_mul_of_one_le_of_one_le ha hb).trans_lt hp1).false
+
+-- ## Step 3: if p does not divide m, then |m|=1
+
+open Real
+
+/-- A natural number not divible by `p` has absolute value 1. -/
+lemma mulRingNorm_eq_one_of_not_dvd {m : ℕ} (hpm : ¬ p ∣ m) : f m = 1 := by
+  apply le_antisymm (bdd m)
+  by_contra! hm
+  set M := f p ⊔ f m with hM
+  set k := Nat.ceil (M.logb (1 / 2)) + 1 with hk
+  obtain ⟨a, b, bezout⟩ : IsCoprime (p ^ k : ℤ) (m ^ k) := by
+    apply IsCoprime.pow (Nat.Coprime.isCoprime _)
+    exact (Nat.Prime.coprime_iff_not_dvd
+      (is_prime_of_minimal_nat_zero_lt_mulRingNorm_lt_one p hp0 hp1 hmin)).2 hpm
+  have le_half {x} (hx0 : 0 < x) (hx1 : x < 1) (hxM : x ≤ M) : x ^ k < 1 / 2 := by
+    calc
+    x ^ k = x ^ (k : ℝ) := (rpow_natCast x k).symm
+    _ < x ^ M.logb (1 / 2) := by
+      apply rpow_lt_rpow_of_exponent_gt hx0 hx1
+      rw [hk]
+      push_cast
+      exact lt_add_of_le_of_pos (Nat.le_ceil (M.logb (1 / 2))) zero_lt_one
+    _ ≤ x ^ x.logb (1 / 2) := by
+      apply rpow_le_rpow_of_exponent_ge hx0 (le_of_lt hx1)
+      simp only [one_div, ← log_div_log, Real.log_inv, neg_div, ← div_neg, hM]
+      gcongr
+      simp only [Left.neg_pos_iff]
+      exact log_neg (lt_sup_iff.2 <| Or.inl hp0) (sup_lt_iff.2 ⟨hp1, hm⟩)
+    _ = 1 / 2 := rpow_logb hx0 (ne_of_lt hx1) one_half_pos
+  apply lt_irrefl (1 : ℝ)
+  calc
+  1 = f 1 := (map_one f).symm
+  _ = f (a * p ^ k + b * m ^ k) := by rw_mod_cast [bezout]; norm_cast
+  _ ≤ f (a * p ^ k) + f (b * m ^ k) := f.add_le' (a * p ^ k) (b * m ^ k)
+  _ ≤ 1 * (f p) ^ k + 1 * (f m) ^ k := by
+    simp only [map_mul, map_pow, le_refl]
+    gcongr
+    all_goals rw [← MulRingNorm.apply_natAbs_eq]; apply bdd
+  _ = (f p) ^ k + (f m) ^ k := by simp only [one_mul]
+  _ < 1 := by
+    have hm₀ : 0 < f m :=
+      map_pos_of_ne_zero _ <| Nat.cast_ne_zero.2 fun H ↦ hpm <| H ▸ dvd_zero p
+    linarith only [le_half hp0 hp1 le_sup_left, le_half hm₀ hm le_sup_right]
+
+-- ## Step 4: |p| = p ^ (- t) for some positive real t
+
+/-- The absolute value of `p` is `p ^ (-t)` for some positive real number `t`. -/
+lemma exists_pos_mulRingNorm_eq_pow_neg : ∃ t : ℝ, 0 < t ∧ f p = p ^ (-t) := by
+  have pprime := is_prime_of_minimal_nat_zero_lt_mulRingNorm_lt_one p hp0 hp1 hmin
+  refine ⟨- logb p (f p), Left.neg_pos_iff.2 <| logb_neg (mod_cast pprime.one_lt) hp0 hp1, ?_⟩
+  rw [neg_neg]
+  refine (rpow_logb (mod_cast pprime.pos) ?_ hp0).symm
+  simp only [ne_eq, Nat.cast_eq_one,Nat.Prime.ne_one pprime, not_false_eq_true]
+
+-- ## Non-archimedean case: end goal
+/--
+  If `f` is bounded and not trivial, then it is equivalent to a p-adic absolute value.
+-/
+theorem mulRingNorm_equiv_padic_of_bounded :
```

Comment:
`p` is unique, right? Is it hard to prove it?

Reply:
That should be quite easy (using that `f(q) = 1` for other primes).

Reply:
I am going to add that! Thanks